README
### Private Audiophile Logs / 오디오파일용 개인 설정 파일

This GitHub is my audiophile configuration and you can modify it as you want.


raspi/Changelog.txt
-. 바뀐 점

21.12 (21-12-01)

01. Raspberry Pi OS 21-10-30 업데이트 (데비안 11 Bullseye, 커널 5.10.63)
02. ALSA 라이브러리 1.2.3.1 업데이트
03. ALSA 설정 파일 튜닝
04. MPD 버퍼 튜닝
05. MPD 상태 파일 메뉴 옵션 추가
06. CPU Affinity 튜닝
07. 셀 redirection 튜닝
08. 커널 파라미터 튜닝
09. consf 설정 스크립트 추가
10. Squeezelite 튜닝
11. Upmpdcli 1.5.13 업데이트 (OpenHome과 UPnP 분리)
12. myMPD 8.1.6 업데이트 (홈 화면 추가)
13. librespot 0.3.1 업데이트
14. HifiBerry DAC+ HD 오디오 카드 추가
15. 웹 설정/인터페이스를 웹 설정으로 통합
16. 클래식 HiFi 세이캐스트와 radioboss.fm 지원
17. 라즈베리 파이 OS 업데이트 사용 안 함 (튜닝과 충돌할 가능성 있음).
18. 시스템 시각 맞추기 (curl 오류 방지)
19. 룬 브리지 업데이트 지원 (roonup)
20. HQPlayer NAA 설치 지원 (naainstall)
21. UPnP 미디어 서버 지원 (gerbera 1.1.0)
22. youtube-dl 대신 yt-dlp 사용

20.06 (20-06-18)

01. ALSA 라이브러리 1.2.3 업데이트.
02. ALSA 설정 튜닝 (min,mix,plug)
03. MPD 버퍼와 믹서 튜닝.
04. MPD 상태 파일 사용 안 함 (오류 발생 시 초기화 안 됨)
05. 커널 스케줄러와 네트워크 설정 튜닝.
06. nohup, qdisc ingress 사용 안 함.
07. 룬 브리지 메모리 직접 실행 (roonbridgetoram)
08. 룬 브리지 디버그 사용 안 함.
09. 룬 브리지 이진 파일 튜닝.
10. 오디오 단축 설정 추가 (audioconf)
11. 라즈베리 파이 OS 2020-05-27 업데이트.
12. Upmpdcli 1.4.10 업데이트와 누락된 타이달 패키지 추가.
13. myMPD 6.4.2 업데이트.
14. 스퀴즈박스 플레이어 추가 (Squeezelite)

20.04.2 (20-04-03 패치) 

1. 중복된 ALSA 설정 수정. 

20.04.1 (20-04-03 패치) 

1. ALSA 설정 튜닝. 

20.04 (20-04-02 패치) 

01. ALSA 설정 튜닝.
02. MPD 믹서와 프로세스 코어 할당 튜닝.

20.03 (20-03-19)

01. ALSA 라이브러리 1.2.2 업데이트.
02. 메모리 라이브러리 경로 추가.
03. 설정 파일 튜닝 (ALSA, MPD, myMPD, upmpdcli)
04. 커널 스케줄러 튜닝.
05. 트래픽 관리와 네트워크 튜닝.
06. 무선랜 모듈 기본 사용 안 함.
07. 라즈비안 업데이트 (2020-02-13일 자)
08. Upmpdcli 1.4.7 버전 업데이트.
09. Upmpdcli Tidal 인증 사용 안 함.
10. Upmpdcli mpd 아이콘 사용 (기존 openwrt 아이콘 어두움)
11. Librespot 0.1.1 업데이트.
12. mympd 6.2.3 업데이트.
13. mympd 메모리 실행.
14. mympd와 ssh 사용 옵션 추가.
15. ttyd 1.6.0 업데이트.
16. OpenHome 라디오 아이콘 일부 바꿈.
17. 클래식HiFi 주소 수정.

19.12 (19-12-12)

01. ALSA 설정 파일 튜닝.
02. 설정 파일 튜닝 (MPD, Upmpdcli, myMPD, Roon)
03. wave와 flac 디코더 빌드.
04. 빌드 파일 디버그 정보 삭제 (ALSA 라이브러리, MPD, librespot, libaudiofile, libFLAC)
05. 모듈과 프로세스 최소화.
06. Upmpdcli Qobuz 메뉴 지움 (더 이상 사용할 수 없음)
07. roontoram 옵션 추가.
08. Spotifyd 대신 Libresot 사용.
09. Spotify 로그인 필요 없음.
10. USB 자동 마운트 오류 수정.
11. 라즈베리 파이 3 전원 LED 밝기 확인.
12. 라즈베리 파이 3 hciuart 사용 안 함.
13. mympd v5.7.2 버전 업데이트.

19.10.1 (19-10-08)

01. airupnp 코어 할당 오류 수정.

19.10 (19-10-08)

01. init와 사운드 출력 스레드 분리.
02. none 블록 스케줄러 적용.
03. ALSA 설정 오버홀.
04. MPD 로그 사용 안 함.
05. MPD 설정 파일 튜닝.
06. 설정 파일 튜닝 (upmpdcli, mympd, roon, spotifyd)
07. 커널 스케줄러 튜닝.
08. audio-config의 '=~ folder' 줄 삭제.
09. mympd 기본 사용자 사용 (USB 음반 표지 지원)
10. mympd 5.6.1 업데이트 (긴 주소 깨짐 수정, http://mympd_ip/library/ 음원 파일 지원)
11. mympd에 오디오 설정(audio-config) 메뉴 추가.
12. 오디오 설정에 리부팅 옵션 추가.
13. SBS 라디오 주소 언패드 추가.
14. 파티션에 PARTUUID 사용.
15. Upmpdcli OpenHome 국내 라디오 추가.
16. Upmpdcli Qobuz, Tidal 메뉴 추가.
17. Z 모뎀 lrzsz 패키지 설치.
18. airupnp 데몬 실행 (안정성과 부팅 속도 개선)

19.08 (19-08-06)

01. 라즈비안 Buster 업그레이드 (라즈베리 파이 4 지원?)
02. MPD 메인과 io 프로세스 변경.
03. Upmpdcli 프로세스 변경.
04. 시스템과 오디오 프로세스 분리 (룬 브리지, 스포티파이, 에어플레이)
05. 프로세스 분리에 따른 ALSA 설정, MPD 버퍼, 커널 스케줄러 튜닝.
06. 룬 브리지 실행 스크립트 튜닝. 
07. tts를 tts2mpd로 대체 (mpd와 구글 번역기 사용)
08. 인터넷 라디오 주소 업데이트 (Audiophile Baroque/Classical/Jazz)
09. myMPD 5.2.2. 한글판 업데이트.
10. hub-ctrl 대신 uhubctl 사용.
11. Audiophonics I-SABRE Q2M DAC, HiFiBerry DAC+ADC 사운드 카드 추가.
12. KBS 라디오 홈페이지의 주소 한 번만 추출.

19.05 (19-05-23)

 1. ALSA 설정 튜닝.
 2. ASLA 더미 믹서 사용.
 3. ALSA 비 호환 모드의 장치 최소화.
 4. ALSA 라이브러리 실행 속성 삭제.
 5. plughw 플러그인 사용을 위한 순정 ALSA 파일 포함.
 6. MPD 버퍼 튜닝.
 7. MPD 상태 파일 사용.
 8. 안정성을 위하여 일부 가상 메모리 설정 사용 안 함.
 9. 커널 스케줄러 조정.
10. SSH 대신 myMPD 웹 클라이언트와 ttyd 웹 터미널 사용.
11. KBS 라디오 주소 변경과 인터넷 라디오 업데이트.
12. 외부 설정 파일 /etc/default/audio.conf 사용.
13. 누락된 TTS 기능 추가.
14. 스포티파이 장치명 변경 지원.
15. 에어플레이 지원.

18.12 (18-11-29)

1. 기본 배포판을 Raspbian에서 DietPi로 바꿈 (안정적인 고해상도 음원 재생)
2. DietPi와 중복되는 audio-config 메뉴 정리.
3. nohup 메모리 재생.
4. 모든 ALSA 라이브러리와 WAV/FLAC 디코더에 메모리 링크 적용.
5. Spotify에 호환 ALSA 라이브러리 적용.
6. Shairport-sync 에어플레이 설치 안 함 (사용자 설치 지원)

18.11 (18-11-15)
 1. 룬 브리지 ALSA 설정의 MPD 출력 장치 명 오류 수정.
 2. 가상 메모리 일부 설정 기본값 적용으로 실행 속도와 안정성 개선.
 3. kr2mpd 스크립트의 KBS, MBC, SBS 라디오 주소 업데이트.
 4. ALSA 설정 추가.
 5. 사용자 더미 믹서 적용.
 6. MPD 0.20.23 버전 업데이트.
 7. MPD mp3 gapless 재생 옵션 지움.
 8. MPD mms 입력 사용 안 함.
 9. MPD oggflac 디코더 사용 (flac 스트림 호환)
10. systemd 코어 할당.
11. 커널 스케줄러 재 설정
12. 사용하지 않는 서비스 로딩 안 함 (부팅 속도 향상과 오류 방지)
13. Raspbian 업데이트.
14. youtube-dl, shairport-sync, spotifyd 버전 업데이트.
15. PiShrink로 이미지를 줄이고 다양한 MicroSD 용량 지원 (leejl529님 추천)

18.07 (18-07-05)
 1. MPD 재생 전 버퍼 설정.
 2. 커널 스케줄러 재 설정.
 3. 기본 CPU Affinity 설정.
 4. exfat 패키지 설치 (exfat-fuse, exfat-utils)
 5. 룬 브리지 추가.
 6. 에어플레이 추가 (Shairport-sync)
 7. Allo Katana DAC(ess9038q2m), JustBoom DAC/Digi, RPi DAC 하드웨어 추가.
 8. Raspbian 업데이트.

18.06 (18-06-14) 
  1. 명령어 실행 널 입출력과 nohup 적용.
  2. MPD 버퍼와 커널 스케줄러 재 설정.

18.05 (18-05-31) 
  1. ALSA 시스템 널 믹서 적용.
  2. MPD와 Upmpdcli 모 프로세스 단순화.
  3. Upmpdcli 전면 배치.
  4. wave와 flac 디코더 메모리 재생.
  5. nohup 사용 안 함.
  6. 명령어 널 입출력 적용과 최적화.
  7. 버퍼 상세 설정.
  8. 커널 스케줄러 재 설정.
  9. Raspbian 업데이트.
10. Spotifyd 추가.

18.04.2~7 (18-05-10)
  1. ALSA 출력 재 설정.

18.04.1 (18-05-01)
  1. 오디오 관련 파일 읽기 전용 설정 되돌리기.
  2. ALSA 출력 재 설정.
  3. MPD 버퍼 상세 설정.
  4. 커널 스케줄러 설정.

18.04 (18-04-26)
  1. ALSA 라이브러리 1.1.5 최소 컴파일.
  2. ALSA 출력은 사용자, 믹서는 사용자+더미 설정.
  3. MPD ALSA 출력 항상 켜기 끔.
  4. MPD 버퍼 상세 설정.
  5. 커널 스케줄러 조정.
  6. Raspbian 업데이트 (신형 라즈베리 파이 3B+ 호환, M2Tech hiFace 지원)
  7. Upmpdcli 1.2.16 업데이트.
  8. 오디오 관련 파일 읽기 전용 설정.
  9. MicroSD 읽기 전용 지원 (remountro, remountrw)

18.03 (18-03-15)
  1. MPD 0.20.18 업그레이드.
  2. MPD mp3 별도 디코더 대신 ffmpeg 사용.
  3. MPD 믹서 사용 안 함.
  4. MPD 출력 우선권 적용 안 함.
  5. ALSA 설정 메모리 로딩.
  6. 가상 메모리 최적화.
  7. 버퍼와 커널 스케줄러 상세 설정.
  8. Raspbian 업데이트.
  9. 인터넷 라디오 개별 .pls 파일 추가.

18.02 (18-02-06)
  1. Raspbian 업데이트.
  2. MPD 0.20.15 업데이트.
  3. MPD와 ALSA 컴파일 옵션 최적화.
  4. MPD의 ALSA 출력과 믹서 재 설정.
  5. MPD 버퍼 상세 설정.
  6. MPD 설정 파일 간략화.
  7. MPD 출력 nice 레벨 조정.
  8. MPD 인터넷 라디오 추가 (MBC, 국악방송, 1.fm Baroque/Blues/Jazz/New Age, Audiophile Jazz, Linn Jazz/Radio)
  9. 시스템 우선권 반영.
10. 부팅 시 ALSA 라이브러리 메모리 링크 갱신.
11. 커널 스케줄러 일부 변경.
12. TTS(Text to Speech) 쉘 스크립트 추가.
13. 메타데이터 선택 메뉴 추가.

kodi/ktvr2kodi.py
#!/usr/bin/python

chs = ["KBS 1FM", "KBS 2FM", "KBS 1R", "KBS 2R", "KBS 3R", "KBS SCR", "KBS RKI",  "KBS 1TV", "KBS 2TV", "KBS WORLD", "KBS24", "KBSN Sports", "KBSN Drama", "KBSN Joy", "KBSN W", "KBSN Life", "KBSN Kids", "MBC Channel M", "MBC FM4U", "MBC FM", "SBS", "SBS Power FM", "SBS Love FM", "SBS Plus", "SBS CNBC", "SBS funE", "SBS Golf", "SBS MTV", "SBS nick", "SBS Sports"]

ch = chs[0]

import urllib, urllib2, re
import xbmc, xbmcgui

def url_func(url):
    WebSock = urllib.urlopen(url)
    WebHTML = WebSock.read()
    WebSock.close()
    return WebHTML

def kbs_func(code):
    Base_URL = 'http://onair.kbs.co.kr/index.html?sname=onair&stype=live&ch_code=' + code
    WebHTML = url_func(Base_URL)
    Temp_Web_URL = re.compile('http://.*_lsu_sa_=[0-9a-z]*').findall(WebHTML)
    url = Temp_Web_URL[0].split('\\', 1)[0]
    ### Double query
    #url_root = url.rsplit('/', 1)[0]
    #M3U = url_func(url)
    #M3U = re.compile('.*m3u.*').findall(M3U)[0]
    #url = url_root + '/' + M3U
    ###
    return url

def mbc_func(code):
    Base_URL = 'http://miniplay.imbc.com/AACLiveURL.ashx?channel=' + code + '&agent=android&protocol=M3U8'
    url = url_func(Base_URL)
    ### Double query
    #url_root = url.rsplit('/', 1)[0]
    #M3U = url_func(url)
    #M3U = re.compile('.*m3u.*').findall(M3U)[0]
    #url = url_root + '/' + M3U
    ###
    return url

def sbs_func(code1, code2):

    Base_URL = 'http://api.sbs.co.kr/vod/_v1/Onair_Media_Auth_Security.jsp?channelPath=' + code1 + '&streamName=' + code2 + '.stream&playerType=mobile'
    data = url_func(Base_URL)

    import base64, Cryptodome.Cipher.DES

    key = b'7d1ff4ea8925c225'
    ciphertext = base64.b64decode(data)
    cipher = Cryptodome.Cipher.DES.new(key[:8], mode=Cryptodome.Cipher.DES.MODE_ECB)
    text = cipher.decrypt(ciphertext)
    url = text.decode('utf8')

    ### Double query
    #print(url)
    #M3U = url_func(url)
    #url = re.compile('.*m3u.*').findall(M3U)[0]
    ###
    return url

def ch_func(ch):
    if   ch == chs[0]:
        url = kbs_func('24')
    elif ch == chs[1]:
        url = kbs_func('25')
    elif ch == chs[2]:
        url = kbs_func('21')
    elif ch == chs[3]:
        url = kbs_func('22')
    elif ch == chs[4]:
        url = kbs_func('23')
    elif ch == chs[5]:
        url = kbs_func('I26')
    elif ch == chs[6]:
        url = kbs_func('I92')
    elif ch == chs[7]:
        url = kbs_func('11')
    elif ch == chs[8]:
        url = kbs_func('12')
    elif ch == chs[9]:
        url = kbs_func('14')
    elif ch == chs[10]:
        url = kbs_func('81')
    elif ch == chs[11]:
        url = kbs_func('N95')
    elif ch == chs[12]:
        url = kbs_func('N91')
    elif ch == chs[13]:
        url = kbs_func('N92')
    elif ch == chs[14]:
        url = kbs_func('N94')
    elif ch == chs[15]:
        url = kbs_func('N93')
    elif ch == chs[16]:
        url = kbs_func('N96')
    elif ch == chs[17]:
        url = mbc_func('chm')
    elif ch == chs[18]:
        url = mbc_func('mfm')
    elif ch == chs[19]:
        url = mbc_func('sfm')
    elif ch == chs[20]:
        url = sbs_func('powerpc', 'powerfm')
    elif ch == chs[21]:
        url = sbs_func('lovepc', 'lovefm')
    elif ch == chs[22]:
        url = sbs_func('sbsch6pc', 'sbsch60')
    elif ch == chs[23]:
        url = sbs_func('sbspluspc', 'sbsplus0')
    elif ch == chs[24]:
        url = sbs_func('sbscnbc', 'sbscnbc0')
    elif ch == chs[25]:
        url = sbs_func('sbsetvpc', 'sbsetv0')
    elif ch == chs[26]:
        url = sbs_func('sbsgolf', 'sbsgolf')
    elif ch == chs[27]:
        url = sbs_func('sbsmtvpc', 'sbsmtv0')
    elif ch == chs[28]:
        url = sbs_func('sbsnickpc', 'sbsnick0')
    elif ch == chs[29]:
        url = sbs_func('sbsespn', 'sbsespn0')
    else:
        print 'Argument(s) is missing or invalid!'
        quit()
    return url

def play_func(ch, url):
    listitem = xbmcgui.ListItem(ch)
    xbmc.Player().play(url, listitem)

url = ch_func(ch)
#print(url)
play_func(ch, url)
